File: 1563-stone-game-v/1563-stone-game-v.py
class Solution:
    def stoneGameV(self, stoneValue: List[int]) -> int:
        @cache
        def dp(left: int, right: int) -> int:
            if left == right:
                return 0
            
            sum_total = 0
            for i in range(left, right + 1):
                sum_total += stoneValue[i]
            
            ans = 0
            sum_left = 0
            for i in range(left, right):
                sum_left += stoneValue[i]
                sum_right = sum_total - sum_left
                if sum_left < sum_right:
                    ans = max(ans, dp(left, i) + sum_left)
                elif sum_left > sum_right:
                    ans = max(ans, dp(i + 1, right) + sum_right)
                else:
                    ans = max(ans, max(dp(left, i), dp(i + 1, right)) + sum_left)
            
            return ans
        
        return dp(0, len(stoneValue) - 1)

# dp top-down + memoization (@cache)  
# time: O(n^3)
# space: O(n^2)

File: 1563-stone-game-v/1563-stone-game-v.go
func stoneGameV(stoneValue []int) int {
    n := len(stoneValue)

    memo := make([][]int, n)
    for i := range memo {
        memo[i] = make([]int, n)
    }

    var dp func(left, right int) int
    dp = func(left, right int) int {
        if left == right {
            return 0
        }

        if memo[left][right] != 0 {
            return memo[left][right]
        }

        sum := 0
        for i := left; i <= right; i++ {
            sum += stoneValue[i]
        }

        sumLeft := 0
        for i := left; i < right; i++ {
            sumLeft += stoneValue[i]
            sumRight := sum - sumLeft
            if sumLeft < sumRight {
                val := dp(left, i) + sumLeft
                if val > memo[left][right] {
                    memo[left][right] = val
                }
            } else if sumLeft > sumRight {
                val := dp(i+1, right) + sumRight
                if val > memo[left][right] {
                    memo[left][right] = val
                }
            } else {
                val := max(dp(left, i), dp(i+1, right)) + sumLeft
                if val > memo[left][right] {
                    memo[left][right] = val
                }
            }
        }

        return memo[left][right]
    }

    return dp(0, n-1)
}

// array, dp top-down + memoization
// time: O(n^3)
// space: O(n^2)